client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java

```diff
@@ -0,0 +1,320 @@
+                        WireCommands.DataAppended dataAppended = (WireCommands.DataAppended) command;
+                        batchSizeTracker.recordAck(dataAppended.getEventNumber());
+                    }
+                    try {
+                        callback.process((Reply) command);
+                    } catch (Exception e) {
+                        callback.processingFailure(e);
+                    }
+                } catch (SocketException e) {
+                    if (e.getMessage().equals("Socket closed")) {
+                        log.info("Closing TcpConnection.Reader because socket is closed.");
+                    } else {
+                        log.warn("Error in reading from socket.", e);
+                    }
+                    stop();
+                } catch (EOFException e) {
+                    log.info("Closing TcpConnection.Reader because end of input readched.");
+                    stop();
+                } catch (Exception e) {
+                    log.warn("Error processing data from from server " + name, e);
+                    stop();
+                }
+            }
+        }
+
+        @VisibleForTesting
+        static WireCommand readCommand(InputStream in, IoBuffer buffer) throws IOException {
+            ByteBuf header = buffer.getBuffOfSize(in, 8);
+
+            int t = header.getInt(0);
+            WireCommandType type = WireCommands.getType(t);
+            if (type == null) {
+                throw new InvalidMessageException("Unknown wire command: " + t);
+            }
+
+            int length = header.getInt(4);
+            if (length < 0 || length > WireCommands.MAX_WIRECOMMAND_SIZE) {
+                throw new InvalidMessageException("Event of invalid length: " + length);
+            }
+
+            ByteBuf payload = buffer.getBuffOfSize(in, length);
+            
+            return type.readFrom(new EnhancedByteBufInputStream(payload), length);
+        }
+        
+        public void stop() {
+            stop.set(true);
+            try {
+                in.close();
+            } catch (Exception e) {
+                log.warn("Got error while shutting down reader {}. ", name, e);
+            }
+            callback.connectionDropped();
+        }
+    }
+    
+    @RequiredArgsConstructor
+    private static final class TimeoutBatch implements Runnable {
+        private final AtomicLong token = new AtomicLong(-1);
+        private final CommandEncoder encoder;
+        @Override
+        public void run() {
+            token.set(encoder.batchTimeout(token.get()));
+        }    
+    }
+
+    public static CompletableFuture<TcpClientConnection> connect(PravegaNodeUri location, ClientConfig clientConfig, ReplyProcessor callback,
+                                              ScheduledExecutorService executor, Runnable onClose) {
+        return CompletableFuture.supplyAsync(() -> {
+            Socket socket = createClientSocket(location, clientConfig); 
```

Comment:
This can throw an exception during socket.connect(...) causing the thread to silently fail.

Reply:
If socket.connect throws the Runnable supplied to `supplyAsync` will fail, which will cause the future returned from connect to fail with the exception. It would only be silent if the caller does not join on the future returned from `connect()`. (Which is unlikely as they need the connection)